src/AppInstallerCommonCore/DODownloader.cpp

```diff
@@ -0,0 +1,395 @@
+                    else if (downloadStatus.State != DODownloadState_Finalized)
+                    {
+                        // For any other state, abort the download since it's no longer in use.
+                        // This will allow DO to clean up the cache for the associated content ID.
+                        (void)LOG_IF_FAILED(m_download->Abort());
+                    }
+                }
+            }
+
+            void SetProperty(DODownloadProperty prop, const std::wstring& value)
+            {
+                wil::unique_variant var;
+                var.bstrVal = ::SysAllocString(value.c_str());
+                THROW_IF_NULL_ALLOC(var.bstrVal);
+                var.vt = VT_BSTR;
+                THROW_IF_FAILED(m_download->SetProperty(prop, &var));
+            }
+
+            void SetProperty(DODownloadProperty prop, std::string_view value)
+            {
+                SetProperty(prop, Utility::ConvertToUTF16(value));
+            }
+
+            void SetProperty(DODownloadProperty prop, uint32_t value)
+            {
+                wil::unique_variant var;
+                var.ulVal = value;
+                var.vt = VT_UI4;
+                THROW_IF_FAILED(m_download->SetProperty(prop, &var));
+            }
+
+            void SetProperty(DODownloadProperty prop, bool value)
+            {
+                wil::unique_variant var;
+                var.boolVal = value ? VARIANT_TRUE : VARIANT_FALSE;
+                var.vt = VT_BOOL;
+                THROW_IF_FAILED(m_download->SetProperty(prop, &var));
+            }
+
+            template<typename T>
+            void SetUnknownProperty(DODownloadProperty prop, T&& value)
+            {
+                wil::unique_variant var;
+                var.punkVal = nullptr;
+                var.vt = VT_UNKNOWN;
+                if (value)
+                {
+                    THROW_IF_FAILED(value->QueryInterface(IID_PPV_ARGS(&var.punkVal)));
+                }
+                THROW_IF_FAILED(m_download->SetProperty(prop, &var));
+            }
+
+            void Uri(std::string_view uri)
+            {
+                SetProperty(DODownloadProperty_Uri, uri);
+            }
+
+            void ContentId(std::string_view contentId)
+            {
+                SetProperty(DODownloadProperty_ContentId, contentId);
+            }
+
+            void DisplayName(std::string_view displayName)
+            {
+                SetProperty(DODownloadProperty_DisplayName, displayName);
+            }
+
+            void LocalPath(const std::filesystem::path& localPath)
+            {
+                SetProperty(DODownloadProperty_LocalPath, localPath.wstring());
+            }
+
+            void CorrelationVector(std::string_view correlationVector)
+            {
+                SetProperty(DODownloadProperty_CorrelationVector, correlationVector);
+            }
+
+            void NoProgressTimeoutSeconds(uint32_t noProgressTimeoutSeconds)

```

Comment:
>NoProgressTimeoutSeconds [](http://example.com/codeflow?start=17&length=24)

Unused. Remove it to avoid someone from introducing a call while still implementing the timeout in DODownloadStatusCallback::Wait?

Reply:
Ah, I see Wait() actually waits indefinitely after the first progress update. Keep in mind DO's default timeouts are 30 minutes for foreground, 3 hours for background.

Are you ok with these defaults?